### PR TITLE
Switch to PackageReference

### LINKED ISSUE #4733: Replacing packages.config with PackageReference
The current `packages.config` approach to listing package dependencies is outdated. Nowadays we could use `PackageReference` instead which comes with a few benefits. From [this page](https://github.com/twsouthwick/NuGetPackageConfigConverter/):

- Transitive dependencies are automatically included - you only need to include the dependencies you need
- install.ps1/uninstal.ps1 is not run so arbitrary code is not run on install and uninstall from untrusted packages
- Doesn't rewrite your project file
- Doesn't add extra files to the project

`PackageReference` would have prevented the error that failed [this build](https://ci.appveyor.com/project/gitextensions/gitextensions/build/3.00.00.2204). In that case a project took an assembly reference with `<HintPath>..\..\..\packages\` which only existed because other projects referenced the package. Once no packages referenced that assembly version any more, the assembly was not available.

I ran that tool over the solution a week ago and it seemed to work, though required a few manual fixes afterwards. I'll try it again now that #4601 is merged.

Does anyone have any experience or thoughts about this?

### PR DESCRIPTION
**Pull request by sharwell**
Closes #4733

### CODE REVIEW

File: GitUI/GitUI.csproj [hunk 1]
```diff
@@ -41,62 +41,30 @@
...
-      <HintPath>..\packages\JetBrains.Annotations.11.1.0\lib\net20\JetBrains.Annotations.dll</HintPath>
+    <Reference Include="ICSharpCode.TextEditor">
+      <HintPath>..\Bin\ICSharpCode.TextEditor.dll</HintPath>
     </Reference>
     <Reference Include="Microsoft.CSharp" />
-    <Reference Include="Microsoft.VisualStudio.Threading, Version=15.6.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\packages\Microsoft.VisualStudio.Threading.15.6.46\lib\net46\Microsoft.VisualStudio.Threading.dll</HintPath>
-      <Private>True</Private>
+    <Reference Include="Microsoft.WindowsAPICodePack">
+      <HintPath>..\Bin\Microsoft.WindowsAPICodePack.dll</HintPath>
```
Comment (RussKie) [inline GitUI/GitUI.csproj]: I think we can replace these with Nuget versions too - https://www.nuget.org/packages/WindowsAPICodePack-Core/

File: GitUI/GitUI.csproj [hunk 2]
```diff
@@ -41,62 +41,30 @@
     <DocumentationFile>bin\Release\GitUI.xml</DocumentationFile>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="JetBrains.Annotations, Version=11.1.0.0, Culture=neutral, PublicKeyToken=1010a0d8d6380325, processorArchitecture=MSIL">
-      <HintPath>..\packages\JetBrains.Annotations.11.1.0\lib\net20\JetBrains.Annotations.dll</HintPath>
+    <Reference Include="ICSharpCode.TextEditor">
+      <HintPath>..\Bin\ICSharpCode.TextEditor.dll</HintPath>
```
Comment (RussKie) [inline GitUI/GitUI.csproj]: This looks like it has been built from this project https://github.com/KindDragon/ICSharpCode.TextEditor

@KindDragon is looks like there was only one functional change made to the code https://github.com/KindDragon/ICSharpCode.TextEditor/pull/3. 
Is there anything that precludes us from referencing the official nuget package https://www.nuget.org/packages/ICSharpCode.TextEditor/?

File: GitUI/GitUI.csproj (hunk 1 above)
Comment (sharwell) [reply GitUI/GitUI.csproj]: ➡️ This would be a separate change (but otherwise in favor of it)

File: GitUI/GitUI.csproj (hunk 2 above)
Comment (sharwell) [reply GitUI/GitUI.csproj]: ➡️ This would be a separate change (but otherwise in favor of it if possible)

File: Directory.Build.props [hunk 3]
```diff
@@ -6,6 +6,11 @@
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <RestoreProjectStyle>PackageReference</RestoreProjectStyle>
+    <RuntimeIdentifiers>win</RuntimeIdentifiers>
```
Comment (sharwell) [inline Directory.Build.props]: :memo: This might not be required anymore, but it doesn't seem to hurt either.

File: Directory.Build.targets [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<Project>
+
+  <ItemGroup Condition="'$(EnableStyleCopAnalyzers)' != 'false'">
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta006" PrivateAssets="all" />
+  </ItemGroup>
+
+  <ItemGroup>
```
Comment (sharwell) [inline Directory.Build.targets]: ❗️ Need to disable this block for the Visual Studio projects or the wrong version will be used.

File: Directory.Build.props (hunk 3 above)
Comment (sharwell) [reply Directory.Build.props]: ➡️ Removed this line

File: Directory.Build.targets (hunk 4 above)
Comment (sharwell) [reply Directory.Build.targets]: ➡️ Fixed and added a comment explaining why this special treatment is needed

File: GitUI/GitUI.csproj (hunk 1 above)
Comment (drewnoakes) [reply GitUI/GitUI.csproj]: #4820

File: GitUI/GitUI.csproj (hunk 2 above)
Comment (KindDragon) [reply GitUI/GitUI.csproj]: > Is there anything that precludes us from referencing the official nuget package https://www.nuget.org/packages/ICSharpCode.TextEditor/?

We used official package before, but we need some patches to make it work for us https://github.com/KindDragon/ICSharpCode.TextEditor/commit/ae3a7fdfc03d5dc3fe424896ef5a60c873930d37

**Comment by gerhardol** [comment]
Time to merge this?

**Comment by RussKie** [comment]
yes

**Comment by sharwell** [comment]
I'll rebase now to fix the conflicts

**Pull request merged**